sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java

```diff
@@ -240,6 +242,35 @@ public void setCoderRegistry(CoderRegistry coderRegistry) {
      * visited.
      */
     public void visitValue(PValue value, TransformTreeNode producer);
+
+    /**
+     * Control enum for indicating whether or not a traversal should process the contents of
+     * a composite transform or not.
+     */
+    public enum Recurse {
+      ENTER_TRANSFORM,
+      DO_NOT_ENTER_TRANSFORM;
+    }
+
+    /**
+     * Default no-op {@link PipelineVisitor} that enters all composite transforms.
+     * User implementations can override just those methods they are interested in.
+     */
+    public class Defaults implements PipelineVisitor {
+      @Override
+      public Recurse enterCompositeTransform(TransformTreeNode node) {
```

Comment:
Actually `visitTransform` is never called for composites. It isn't obvious how it should be interleaved with the visiting of its children, to me.

We should probably take the opportunity to rename `visitTransform` to something that makes clear it isn't for all transforms, just non-composites.